lib/SILOptimizer/Analysis/ArraySemantic.cpp

```diff
@@ -376,7 +386,27 @@ static SILValue copyArrayLoad(SILValue ArrayStructValue,
     InsertPt = Inst;
   }
 
-  return cast<LoadInst>(LI->clone(InsertBefore));
+  if (ArrayStructValue->getFunction()->hasOwnership()) {
+    if (LeaveOriginal) {
+      if (auto *Load = dyn_cast<LoadInst>(ArrayStructValue)) {
+        return Load->clone(InsertBefore);
+      }
+      auto *Copy = cast<CopyValueInst>(ArrayStructValue);
+      auto Addr = cast<LoadInst>(Copy->getOperand())->getOperand();
+      return SILBuilderWithScope(InsertPt).createLoad(
+          InsertPt->getLoc(), Addr, LoadOwnershipQualifier::Copy);
+    }
+    if (auto *Load = dyn_cast<LoadInst>(ArrayStructValue)) {
+      Load->moveBefore(InsertBefore);
+      return ArrayStructValue;
+    }
+    auto *Copy = cast<CopyValueInst>(ArrayStructValue);
+    auto Addr = cast<LoadInst>(Copy->getOperand())->getOperand();
+    Copy->replaceAllUsesWithUndef();
```

Comment:
I can't understand how `replaceAllUsesWithUndef` works here. What about other uses of the array value? The call may take a `guaranteed` self, or may be in a different block from the copy, no?

Reply:
Fixed it.